### PR TITLE
Add Azure Application Insights connection string support

### LINKED ISSUE #3710: Migrate from Application Insights instrumentation keys to connection strings
> On 31 March 2025, technical support for instrumentation key–based global ingestion in the Application Insights feature of Azure Monitor will end.

How can we use connection strings with Micrometer?

https://github.com/micrometer-metrics/micrometer/blob/a350afa3cff4e6df1a68f394250affc39620a4e7/implementations/micrometer-registry-azure-monitor/src/main/java/io/micrometer/azuremonitor/AzureMonitorConfig.java#L42-L44

https://azure.microsoft.com/en-us/updates/technical-support-for-instrumentation-key-based-global-ingestion-in-application-insights-will-end-on-31-march-2025/

Thanks,
Christophe

### PR DESCRIPTION
**Pull request by cbismuth**
This implementation overrides instrumentation key when connection string is set from the `AzureMonitorConfig` class.

Fixes #3710.

### CODE REVIEW

File: implementations/micrometer-registry-azure-monitor/src/main/java/io/micrometer/azuremonitor/AzureMonitorMeterRegistry.java
Comment (shakuzen) [inline implementations/micrometer-registry-azure-monitor/src/main/java/io/micrometer/azuremonitor/AzureMonitorMeterRegistry.java:68]: If you just configure the connection string (containing the instrumentation key) but do not configure the `instrumentationKey` on `AzureMonitorConfig`, won't this throw an error here? It seems like we should check first for and apply the connection string.

**Review by shakuzen — CHANGES_REQUESTED** [review]

File: implementations/micrometer-registry-azure-monitor/src/main/java/io/micrometer/azuremonitor/AzureMonitorUtils.java
Comment (cbismuth) [reply implementations/micrometer-registry-azure-monitor/src/main/java/io/micrometer/azuremonitor/AzureMonitorUtils.java]: I agree, yes.

File: implementations/micrometer-registry-azure-monitor/src/main/java/io/micrometer/azuremonitor/AzureMonitorMeterRegistry.java
Comment (cbismuth) [reply implementations/micrometer-registry-azure-monitor/src/main/java/io/micrometer/azuremonitor/AzureMonitorMeterRegistry.java]: I don't think an error would be thrown, but I pushed a new proposal relying on connection string only.

**Comment by cbismuth** [comment]
Hi @shakuzen, here is a new proposal with instrumentation key deprecation and connection string as the new default implementation, no backward compatibility breakage. What do you think of this? thanks.

**Review by shakuzen — CHANGES_REQUESTED** [review]
Thank you for the updated changes. I think we're moving in the right direction. While things are API compatible with the changes as is, a user upgrading to these changes will not be able to use this without making changes to their configuration. I think we can provide a smoother and more gradual upgrade path. What do you think of the following idea? If the user sets connectionString, we use that. If the user does not set connectionString but does set instrumentationKey (users upgrading to the version with these changes will be in this situation initially), we will by default return from connectionString `InstrumentationKey=` concatenated with the instrumentationKey. If neither are set, then a validation error should be thrown. We should have unit tests for these different cases. In order to prove the easy upgrade path for users, all the existing tests should pass as-is. We can add additional tests for new behavior.

File: implementations/micrometer-registry-azure-monitor/src/main/java/io/micrometer/azuremonitor/AzureMonitorMeterRegistry.java
Comment (cbismuth) [reply implementations/micrometer-registry-azure-monitor/src/main/java/io/micrometer/azuremonitor/AzureMonitorMeterRegistry.java:72]: @sonatype-lift ignore

**Comment by cbismuth** [comment]
Yes you're right @shakuzen, this is far more readable, thanks for pointing this.

**Review by shakuzen — APPROVED** [review]
Thank you for the pull request and being so responsive to feedback. We're ready to merge this. I will polish some things in a follow-up commit after merging. Have you had a chance to try this out with an application just to verify the new configuration is working with sending metrics to Application Insights? If not, please do try out snapshots when they are published post-merge. Thank you again!

**Comment by shakuzen** [comment]
I don't know why the CI build is having trouble running, but I ran the build locally on this pull request and it passed, so I'll merge anyway.

**Comment by cbismuth** [comment]
Thanks @shakuzen! I didn't manage to make Circle CI build work on my side neither. I'll try this locally by the end of the week, no worries, I'll let you know, have a nice day.

**Pull request merged**
